wd_mempool.c

```diff
@@ -0,0 +1,105 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+#include <ctype.h>
+#include <dirent.h>
+#include <errno.h>
+#include <poll.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include "wd.h"
+
+struct blkpool {
+	void **blk_elem; /* all the block unit addrs saved in blk_elem */
+	size_t depth;    /* the block pool deph, stack depth */
+	size_t top;      /* the stack top pos for blk_elem */
+	size_t blk_size; /* the size of one block */
+	handle_t mp;     /* record from which mempool */
+};
+
+struct mempool {
+	void *addr;
+	size_t size;
+};
+
+void *wd_blockpool_alloc(handle_t blockpool)
+{
+	struct blkpool *bp = (struct blkpool*)blockpool;
+	
+	if (!bp)
+		return NULL;
+
+	if (bp->top > 0) {
+		bp->top--;
+		return bp->blk_elem[bp->top];
+	}
+	
+	return NULL;
+}
+
+void wd_blockpool_free(handle_t blockpool, void *addr)
+{
+	struct blkpool *bp = (struct blkpool*)blockpool;
+	
+	if (!bp || !addr)
+		return;
+
+	if (bp->top < bp->depth) {
+		bp->blk_elem[bp->top] = addr;
```

Comment:
If there're four blocks, all these four blocks (block 0-3) are allocated by wd_blockpool_alloc(). Now block 3 is freed by wd_blockpool_free().
At this time, bp->top is 0. The address of block 3 will be filled into block 0 since "top" field is still 0.

And why should we update "blk_elem" field at here? It seems that nobody will clean it.